src/interpolation/details/ArborX_InterpDetailsSymmetricPseudoInverseSVD.hpp

```diff
@@ -0,0 +1,291 @@
+                "Each input matrix must have the same value type");
+  KOKKOS_ASSERT(A.extent(0) == ES.extent(0) && ES.extent(0) == U.extent(0));
+  using value_t = typename AMatrix::non_const_value_type;
+  int const size = A.extent(0);
+
+  // We first initialize U as the identity matrix and copy A to ES
+  for (int i = 0; i < size; i++)
+    for (int j = 0; j < size; j++)
+    {
+      U(i, j) = value_t(i == j);
+      ES(i, j) = A(i, j);
+    }
+
+  static constexpr value_t epsilon = Kokkos::Experimental::epsilon_v<float>;
+  while (true)
+  {
+    // We have a guarantee that p < q
+    auto const [max_val, p, q] = argmaxUpperTriangle(ES);
+    if (max_val <= epsilon)
+      break;
+
+    auto const a = ES(p, p);
+    auto const b = ES(p, q);
+    auto const c = ES(q, q);
+
+    // Our submatrix is now
+    // +----------+----------+   +---+---+
+    // | ES(p, p) | ES(p, q) |   | a | b |
+    // +----------+----------+ = +---+---+
+    // | ES(q, p) | ES(q, q) |   | b | c |
+    // +----------+----------+   +---+---+
+
+    // Lets compute x, y and theta such that
+    // +---+---+              +---+---+
+    // | a | b |              | x | 0 |
+    // +---+---+ = R(theta) * +---+---+ * R(theta)^T
+    // | b | c |              | 0 | y |
+    // +---+---+              +---+---+
+
+    value_t cos_theta, sin_theta, x, y;
+    if (a == c)
+    {
+      cos_theta = Kokkos::sqrt(value_t(2)) / 2;
+      sin_theta = cos_theta;
+      x = a + b;
+      y = a - b;
+    }
+    else
+    {
+      auto const u = (2 * b) / (a - c);
+      auto const v = 1 / Kokkos::sqrt(u * u + 1);
+      cos_theta = Kokkos::sqrt((1 + v) / 2);
+      sin_theta = Kokkos::copysign(Kokkos::sqrt((1 - v) / 2), u);
+      x = (a + c + (a - c) / v) / 2;
+      y = a + c - x;
+    }
+
+    // Now lets compute the following new values for U and ES
+    // ES <- R'(theta)^T . ES . R'(theta)
+    // U  <- U . R'(theta)
+
+    // R'(theta)^T . ES . R'(theta)
+    int i = 0;
+    for (; i < p; i++)
+    {
+      auto const es_ip = ES(i, p);
+      auto const es_iq = ES(i, q);
+      ES(i, p) = cos_theta * es_ip + sin_theta * es_iq;
+      ES(i, q) = -sin_theta * es_ip + cos_theta * es_iq;
+    }
+    ES(p, p) = x;
+    i++;
+    for (; i < q; i++)
+    {
+      auto const es_pi = ES(p, i);
+      auto const es_iq = ES(i, q);
+      ES(p, i) = cos_theta * es_pi + sin_theta * es_iq;
+      ES(i, q) = -sin_theta * es_pi + cos_theta * es_iq;
+    }
+    ES(q, q) = y;
+    i++;
+    for (; i < size; i++)
+    {
+      auto const es_pi = ES(p, i);
+      auto const es_qi = ES(q, i);
+      ES(p, i) = cos_theta * es_pi + sin_theta * es_qi;
+      ES(q, i) = -sin_theta * es_pi + cos_theta * es_qi;
+    }
+    ES(p, q) = 0;
+
+    // U . R'(theta)
+    for (int i = 0; i < size; i++)
+    {
+      auto const u_ip = U(i, p);
+      auto const u_iq = U(i, q);
+      U(i, p) = cos_theta * u_ip + sin_theta * u_iq;
+      U(i, q) = -sin_theta * u_ip + cos_theta * u_iq;
+    }
```

Comment:
Would it make sense to use hierarchical parallelism for these for-loops or do we assume `size` to be small enough that it doesn't matter? Or do we have sufficient parallelism already in the outer loop?

Reply:
It would make a lot of sense to use hierarchical parallelism here (and in other loops) as well as using scratch pads for the auxiliary matrices. However, because this is part of the MLS algorithm, it might not be a bottleneck and it is best to avoid early optimization.

However, the following should be a logical distribution:
- Each team treats a single matrix
- Threads/Vectors handle the loops and reductions
- Use the team scratch pad for `ES` and `U`

Reply:
Fair enough.